File: scripts/execute-schema.py
#!/usr/bin/env python3
"""
Execute database schema SQL against Supabase or PostgreSQL
Usage: python3 execute-schema.py <supabase_url> <service_role_key>
"""

import sys
import os
import json
from pathlib import Path

def execute_schema():
    """Execute the schema SQL file"""

    # Get Supabase credentials from arguments or environment
    supabase_url = os.getenv('SUPABASE_URL') or (sys.argv[1] if len(sys.argv) > 1 else None)
    service_role_key = os.getenv('SUPABASE_SERVICE_ROLE_KEY') or (sys.argv[2] if len(sys.argv) > 2 else None)

    if not supabase_url or not service_role_key:
        print("❌ Error: Missing Supabase credentials")
        print("\nUsage:")
        print("  python3 execute-schema.py <SUPABASE_URL> <SERVICE_ROLE_KEY>")
        print("\nOr set environment variables:")
        print("  export SUPABASE_URL=https://your-project.supabase.co")
        print("  export SUPABASE_SERVICE_ROLE_KEY=your-key")
        sys.exit(1)

    # Read schema file
    schema_path = Path(__file__).parent.parent / 'src' / 'db' / 'schema.sql'

    if not schema_path.exists():
        print(f"❌ Schema file not found: {schema_path}")
        sys.exit(1)

    with open(schema_path, 'r') as f:
        schema_sql = f.read()

    print("🔄 Initializing database schema...")
    print(f"📍 Connecting to: {supabase_url}")
    print(f"\n📋 Schema file size: {len(schema_sql)} bytes")

    # Try using psycopg2 (direct PostgreSQL)
    try:
        import psycopg2
        from urllib.parse import urlparse

        # Parse Supabase connection string
        # Supabase format: https://project-id.supabase.co
        # We need to construct PostgreSQL connection URL

        print("\n✅ Using psycopg2 for direct PostgreSQL connection")

        # For Supabase, we need the direct PostgreSQL connection
        # URL format: postgresql://user:password@host/database
        # You can get this from Supabase: Project Settings → Database → Connection string

        db_url = os.getenv('DATABASE_URL')
        if not db_url:
            print("\n⚠️  DATABASE_URL not set. Cannot use direct PostgreSQL connection.")
            print("   Please use the Supabase web UI to execute the schema:")
            print("   1. Open: https://app.supabase.com/project/[project-id]/sql/new")
            print("   2. Paste the contents of src/db/schema.sql")
            print("   3. Click 'Run' → 'Run without RLS'")
            sys.exit(1)

        conn = psycopg2.connect(db_url, sslmode='require')
        cursor = conn.cursor()

        print("\n📝 Executing schema SQL...\n")

        # Execute the entire schema
        cursor.execute(schema_sql)
        conn.commit()

        print("✅ Schema executed successfully!")

        # Verify tables
        cursor.execute("""
            SELECT table_name
            FROM information_schema.tables
            WHERE table_schema = 'public'
            ORDER BY table_name
        """)

        tables = cursor.fetchall()
        print("\n📋 Created tables:")
        for (table_name,) in tables:
            print(f"   ✅ {table_name}")

        cursor.close()
        conn.close()

        print("\n✨ Database initialization complete!")
        return 0

    except ImportError:
        print("\n⚠️  psycopg2 not available. Trying alternative methods...\n")
    except Exception as e:
        print(f"\n❌ Error: {e}\n")

    # Fallback: Instructions for using Supabase web UI
    print("📝 To execute the schema, use the Supabase web interface:\n")
    print("1. Open your Supabase project:")
    print(f"   https://app.supabase.com/project/[project-id]/sql/new\n")
    print("2. Copy the schema SQL from: src/db/schema.sql\n")
    print("3. Paste it into the SQL Editor query box\n")
    print("4. Click 'Run' button (or Cmd+Enter)\n")
    print("5. When prompted about RLS, select 'Run without RLS'\n")
    print("6. Verify the 4 tables were created in Table Editor\n")

    return 1

if __name__ == '__main__':
    sys.exit(execute_schema())

File: scripts/execute-schema.js
#!/usr/bin/env node

/**
 * Execute schema SQL directly against Supabase
 * This bypasses the web UI and executes SQL statements via the JavaScript client
 */

const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const path = require('path');

// Get credentials from environment or arguments
const SUPABASE_URL = process.env.SUPABASE_URL || process.argv[2];
const SUPABASE_SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY || process.argv[3];

if (!SUPABASE_URL || !SUPABASE_SERVICE_ROLE_KEY) {
  console.error('❌ Error: Missing Supabase credentials');
  console.error('Usage: node execute-schema.js <SUPABASE_URL> <SUPABASE_SERVICE_ROLE_KEY>');
  console.error('Or set SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY environment variables');
  process.exit(1);
}

// Create Supabase client with service role key
const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY);

async function executeSchema() {
  try {
    console.log('🔄 Initializing database schema...');
    console.log(`📍 Connecting to: ${SUPABASE_URL}`);

    // Read the schema SQL file
    const schemaPath = path.join(__dirname, '../src/db/schema.sql');
    const schemaSql = fs.readFileSync(schemaPath, 'utf8');

    console.log('\n📝 Schema SQL file loaded successfully');
    console.log(`   File size: ${schemaSql.length} bytes`);

    // Split into individual statements (simple approach - may need refinement for complex SQL)
    const statements = schemaSql
      .split(';')
      .map(stmt => stmt.trim())
      .filter(stmt => stmt.length > 0 && !stmt.startsWith('--'));

    console.log(`\n📋 Found ${statements.length} SQL statements to execute`);

    // Execute statements one by one
    let successCount = 0;
    let errorCount = 0;

    for (let i = 0; i < statements.length; i++) {
      const statement = statements[i];
      const statementNum = i + 1;

      // Show progress for larger statements
      if (statement.length > 100) {
        const preview = statement.substring(0, 60).replace(/\n/g, ' ') + '...';
        console.log(`\n[${statementNum}/${statements.length}] Executing: ${preview}`);
      } else {
        console.log(`\n[${statementNum}/${statements.length}] Executing: ${statement.replace(/\n/g, ' ')}`);
      }

      try {
        // Use the admin API to execute raw SQL
        const { data, error } = await supabase.rpc('exec', {
          sql: statement
        });

        if (error) {
          // If rpc doesn't work, the error will be caught below
          throw error;
        }

        console.log(`   ✅ Success`);
        successCount++;
      } catch (err) {
        // Some statements might fail if they depend on extensions or previous statements
        // This is normal - we'll continue and check if tables exist at the end
        if (err.message.includes('does not exist') || err.message.includes('already exists')) {
          console.log(`   ⚠️  Note: ${err.message}`);
        } else {
          console.log(`   ❌ Error: ${err.message}`);
          errorCount++;
        }
      }
    }

    console.log(`\n📊 Execution Summary:`);
    console.log(`   ✅ Successful: ${successCount}`);
    console.log(`   ❌ Errors: ${errorCount}`);

    // Verify tables exist
    console.log('\n🔍 Verifying tables were created...');

    const tablesToCheck = ['outages', 'outage_history', 'postcode_cache', 'data_fetch_log'];
    let allTablesExist = true;

    for (const tableName of tablesToCheck) {
      try {
        const { data, error } = await supabase
          .from(tableName)
          .select('count', { count: 'exact', head: true });

        if (error && error.code === 'PGRST116') {
          console.log(`   ❌ ${tableName}: Does not exist`);
          allTablesExist = false;
        } else if (error) {
          console.log(`   ⚠️  ${tableName}: ${error.message}`);
        } else {
          console.log(`   ✅ ${tableName}: Exists`);
        }
      } catch (err) {
        console.log(`   ⚠️  ${tableName}: Could not verify - ${err.message}`);
      }
    }

    if (allTablesExist) {
      console.log('\n✨ Database schema initialized successfully!');
      console.log('\n📋 Next steps:');
      console.log('1. Test Node.js connection: node src/db/test-connection.js');
      console.log('2. Build UKPN API data fetcher');
      console.log('3. Implement web scrapers for other DNOs');
      console.log('4. Set up cron job for periodic updates');
      process.exit(0);
    } else {
      console.log('\n⚠️  Some tables may not have been created');
      console.log('This might be due to RLS settings or extension issues.');
      console.log('Try executing the schema via Supabase SQL Editor instead:');
      console.log('1. Open https://app.supabase.com/project/[your-project-id]/sql/new');
      console.log('2. Paste the contents of src/db/schema.sql');
      console.log('3. Click "Run" and select "Run without RLS"');
      process.exit(1);
    }
  } catch (err) {
    console.error('\n❌ Fatal error:', err.message);
    process.exit(1);
  }
}

// Run the execution
executeSchema();
